File: tests/cuda/julia/test_cuda_basic.jl
# CUDA binding tests — 35 tests aligned with CPU Python/Lua/Julia
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_cuda_basic.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

# ============================================================================
# Creation (7 tests)
# ============================================================================
println("=== Creation ===")

a = Insight.zeros([2, 3], Insight.float32)
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("zeros", Insight.numel(a_gpu) == 6)

a = Insight.ones([2, 3], Insight.float32)
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("ones", Insight.numel(a_gpu) == 6)

a = Insight.full([2, 3], 7.0, Insight.float32)
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("full", Insight.numel(a_gpu) == 6)

a = Insight.eye(3)
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("eye", Insight.numel(a_gpu) == 9)

a = Insight.arange(10, Insight.float32)
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("arange", Insight.numel(a_gpu) == 10)

a = Insight.linspace(0.0, 1.0, 5, Insight.float64)
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("linspace", Insight.numel(a_gpu) == 5)

a = Insight.from_data([1.5, 2.5, 3.5])
a_gpu = Insight.to_array(a, Insight.GPUPlace(0))
check("from_data", Insight.numel(a_gpu) == 3)

# ============================================================================
# Arithmetic (5 tests)
# ============================================================================
println("=== Arithmetic ===")

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
b = Insight.to_array(Insight.from_data([4.0, 5.0, 6.0]), Insight.GPUPlace(0))

c = Insight.add(a, b)
check("add", Insight.numel(c) == 3)

c = Insight.sub(a, b)
check("sub", Insight.numel(c) == 3)

c = Insight.mul(a, b)
check("mul", Insight.numel(c) == 3)

c = Insight.div(a, b)
check("div", Insight.numel(c) == 3)

c = Insight.negative(a)
check("neg", Insight.numel(c) == 3)

# ============================================================================
# Unary (5 tests)
# ============================================================================
println("=== Unary ===")

a = Insight.to_array(Insight.from_data([-3.0, -1.0, 0.0, 2.0, 4.0]), Insight.GPUPlace(0))
b = Insight.abs(a)
check("abs", Insight.numel(b) == 5)

a = Insight.to_array(Insight.from_data([1.0, 4.0, 9.0, 16.0]), Insight.GPUPlace(0))
b = Insight.sqrt(a)
check("sqrt", Insight.numel(b) == 4)

a = Insight.to_array(Insight.from_data([0.0, 1.0, 2.0]), Insight.GPUPlace(0))
b = Insight.exp(a)
check("exp", Insight.numel(b) == 3)

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
b = Insight.log(a)
check("log", Insight.numel(b) == 3)

a = Insight.to_array(Insight.from_data([0.0, 0.5, 1.0]), Insight.GPUPlace(0))
b = Insight.sin(a)
check("sin", Insight.numel(b) == 3)

# ============================================================================
# Reduction (5 tests)
# ============================================================================
println("=== Reduction ===")

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0]), Insight.GPUPlace(0))

s = Insight.sum(a)
check("sum", Insight.numel(s) == 1)

m = Insight.mean(a)
check("mean", Insight.numel(m) == 1)

a = Insight.to_array(Insight.from_data([3.0, 1.0, 4.0, 1.0, 5.0]), Insight.GPUPlace(0))
m = Insight.max(a)
check("max", Insight.numel(m) == 1)

m = Insight.min(a)
check("min", Insight.numel(m) == 1)

a = Insight.to_array(Insight.from_data([1.0, 5.0, 3.0, 2.0]), Insight.GPUPlace(0))
m = Insight.argmax(a)
check("argmax", Insight.numel(m) == 1)

# ============================================================================
# Comparison (4 tests)
# ============================================================================
println("=== Comparison ===")

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
b = Insight.to_array(Insight.from_data([1.0, 0.0, 3.0]), Insight.GPUPlace(0))
c = Insight.equal(a, b)
check("equal", Insight.numel(c) == 3)

a = Insight.to_array(Insight.from_data([3.0, 1.0, 5.0]), Insight.GPUPlace(0))
b = Insight.to_array(Insight.from_data([2.0, 4.0, 5.0]), Insight.GPUPlace(0))
c = Insight.greater(a, b)
check("greater", Insight.numel(c) == 3)

c = Insight.less(a, b)
check("less", Insight.numel(c) == 3)

a_bool = Insight.to_array(Insight.from_data([1, 1, 0], Insight.bool), Insight.GPUPlace(0))
b_bool = Insight.to_array(Insight.from_data([1, 0, 0], Insight.bool), Insight.GPUPlace(0))
c = Insight.logical_and(a_bool, b_bool)
check("logical_and", Insight.numel(c) == 3)

# ============================================================================
# Manipulation (3 tests)
# ============================================================================
println("=== Manipulation ===")

a = Insight.to_array(Insight.arange(6, Insight.float64), Insight.GPUPlace(0))
b = Insight.reshape(a, [2, 3])
check("reshape", Insight.numel(b) == 6)

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0, 6.0]), Insight.GPUPlace(0))
a = Insight.reshape(a, [2, 3])
b = Insight.transpose(a)
check("transpose", Insight.numel(b) == 6)

a = Insight.to_array(Insight.zeros([1, 3, 1], Insight.float64), Insight.GPUPlace(0))
b = Insight.squeeze(a)
check("squeeze", Insight.numel(b) == 3)

# ============================================================================
# Linalg (3 tests)
# ============================================================================
println("=== Linalg ===")

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0]), Insight.GPUPlace(0))
a = Insight.reshape(a, [2, 2])
b = Insight.to_array(Insight.from_data([5.0, 6.0, 7.0, 8.0]), Insight.GPUPlace(0))
b = Insight.reshape(b, [2, 2])

c = Insight.matmul(a, b)
check("matmul", Insight.numel(c) == 4)

try
    local d = Insight.det(a)
    check("det", Insight.numel(d) == 1)
catch e
    println("SKIP: det (requires cuBLAS)")
end

try
    local b = Insight.inv(a)
    check("inv", Insight.numel(b) == 4)
catch e
    println("SKIP: inv (requires cuBLAS)")
end

# ============================================================================
# FFT (2 tests)
# ============================================================================
println("=== FFT ===")

x = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0]), Insight.GPUPlace(0))
y = Insight.fft(x)
check("fft", Insight.numel(y) == 4)

z = Insight.ifft(y)
check("ifft", Insight.numel(z) == 4)

# ============================================================================
# Signal (1 test)
# ============================================================================
println("=== Signal ===")

w = Insight.signal.hann(16)
check("hann", Insight.numel(w) == 16)

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end

File: tests/cuda/python/test_cuda_basic.py
"""CUDA binding tests — 35 tests aligned with CPU Python/Lua/Julia.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python \\
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \\
    python -m pytest tests/cuda/python/test_cuda_basic.py -v
"""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr, dtype=None):
    if dtype is not None:
        np_arr = np_arr.astype(dtype)
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.contiguous().to(CPU).numpy()


# ============================================================================
# Creation (7 tests)
# ============================================================================


class TestCreation:
    """Array creation on GPU — tests 1-7."""

    def test_zeros(self):
        a = ins.zeros([2, 3], ins.float32).to(GPU)
        assert a.numel == 6
        assert np.allclose(to_numpy(a), 0)

    def test_ones(self):
        a = ins.ones([2, 3], ins.float32).to(GPU)
        assert a.numel == 6
        np.testing.assert_allclose(to_numpy(a), np.ones([2, 3]))

    def test_full(self):
        a = ins.full([2, 3], 7.0, ins.float32).to(GPU)
        assert a.numel == 6
        np.testing.assert_allclose(to_numpy(a), np.full([2, 3], 7.0))

    def test_eye(self):
        a = ins.eye(3).to(GPU)
        assert a.numel == 9
        np.testing.assert_allclose(to_numpy(a), np.eye(3))

    def test_arange(self):
        a = ins.arange(10, ins.float32).to(GPU)
        assert a.numel == 10
        np.testing.assert_allclose(to_numpy(a), np.arange(10, dtype=np.float32))

    def test_linspace(self):
        a = ins.linspace(0.0, 1.0, 5, ins.float64).to(GPU)
        assert a.numel == 5
        np.testing.assert_allclose(to_numpy(a), np.linspace(0, 1, 5), rtol=1e-6)

    def test_from_numpy(self):
        data = np.array([1.5, 2.5, 3.5], dtype=np.float64)
        a = ins.from_numpy(data).to(GPU)
        assert a.numel == 3
        np.testing.assert_allclose(to_numpy(a), data)


# ============================================================================
# Arithmetic (5 tests)
# ============================================================================


class TestArithmetic:
    """Elementwise arithmetic on GPU — tests 8-12."""

    def test_add(self):
        a_np = np.array([1, 2, 3], dtype=np.float64)
        b_np = np.array([4, 5, 6], dtype=np.float64)
        result = ins.add(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np + b_np, rtol=1e-10)

    def test_sub(self):
        a_np = np.array([10, 20, 30], dtype=np.float64)
        b_np = np.array([1, 2, 3], dtype=np.float64)
        result = ins.sub(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np - b_np, rtol=1e-10)

    def test_mul(self):
        a_np = np.array([2, 3, 4], dtype=np.float64)
        b_np = np.array([5, 6, 7], dtype=np.float64)
        result = ins.mul(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np * b_np, rtol=1e-10)

    def test_div(self):
        a_np = np.array([10, 20, 30], dtype=np.float64)
        b_np = np.array([2, 4, 5], dtype=np.float64)
        result = ins.div(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np / b_np, rtol=1e-10)

    def test_neg(self):
        a_np = np.array([1, -2, 3], dtype=np.float64)
        result = ins.negative(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), -a_np, rtol=1e-10)


# ============================================================================
# Unary (5 tests)
# ============================================================================


class TestUnary:
    """Unary math on GPU — tests 13-17."""

    def test_abs(self):
        a_np = np.array([-3, -1, 0, 2, 4], dtype=np.float64)
        result = ins.abs(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.abs(a_np), rtol=1e-10)

    def test_sqrt(self):
        a_np = np.array([1, 4, 9, 16], dtype=np.float64)
        result = ins.sqrt(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.sqrt(a_np), rtol=1e-10)

    def test_exp(self):
        a_np = np.array([0, 1, 2], dtype=np.float64)
        result = ins.exp(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.exp(a_np), rtol=1e-10)

    def test_log(self):
        a_np = np.array([1, 2, 3], dtype=np.float64)
        result = ins.log(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.log(a_np), rtol=1e-10)

    def test_sin(self):
        a_np = np.array([0, 0.5, 1.0], dtype=np.float64)
        result = ins.sin(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.sin(a_np), rtol=1e-10)


# ============================================================================
# Reduction (5 tests)
# ============================================================================


class TestReduction:
    """Reduction ops on GPU — tests 18-22."""

    def test_sum(self):
        a = to_gpu(np.array([1, 2, 3, 4], dtype=np.float64))
        result = ins.sum(a)
        assert result.numpy().item() == pytest.approx(10.0)

    def test_mean(self):
        a = to_gpu(np.array([1, 2, 3, 4], dtype=np.float64))
        result = ins.mean(a)
        assert result.numpy().item() == pytest.approx(2.5)

    def test_max(self):
        a = to_gpu(np.array([3, 1, 4, 1, 5], dtype=np.float64))
        result = ins.max(a)
        assert result.numpy().item() == pytest.approx(5.0)

    def test_min(self):
        a = to_gpu(np.array([3, 1, 4, 1, 5], dtype=np.float64))
        result = ins.min(a)
        assert result.numpy().item() == pytest.approx(1.0)

    def test_argmax(self):
        a = to_gpu(np.array([1, 5, 3, 2], dtype=np.float64))
        result = ins.argmax(a)
        assert result.numpy().item() == 1


# ============================================================================
# Comparison (4 tests)
# ============================================================================


class TestComparison:
    """Comparison ops on GPU — tests 23-26."""

    def test_equal(self):
        a = to_gpu(np.array([1, 2, 3], dtype=np.float64))
        b = to_gpu(np.array([1, 0, 3], dtype=np.float64))
        result = ins.equal(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, True]))

    def test_greater(self):
        a = to_gpu(np.array([3, 1, 5], dtype=np.float64))
        b = to_gpu(np.array([2, 4, 5], dtype=np.float64))
        result = ins.greater(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, False]))

    def test_less(self):
        a = to_gpu(np.array([1, 5, 3], dtype=np.float64))
        b = to_gpu(np.array([2, 4, 5], dtype=np.float64))
        result = ins.less(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, True]))

    def test_logical_and(self):
        a = to_gpu(np.array([True, True, False], dtype=np.bool_))
        b = to_gpu(np.array([True, False, False], dtype=np.bool_))
        result = ins.logical_and(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, False]))


# ============================================================================
# Manipulation (3 tests)
# ============================================================================


class TestManipulation:
    """Manipulation ops on GPU — tests 27-29."""

    def test_reshape(self):
        a = to_gpu(np.arange(6, dtype=np.float64))
        b = ins.reshape(a, [2, 3])
        assert b.numel == 6

    def test_transpose(self):
        a = to_gpu(np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64))
        b = ins.permute(a, [1, 0])
        np.testing.assert_allclose(
            to_numpy(b), np.array([[1, 4], [2, 5], [3, 6]], dtype=np.float64)
        )

    def test_squeeze(self):
        a = to_gpu(np.zeros([1, 3, 1], dtype=np.float64))
        b = ins.squeeze(a)
        assert b.numel == 3


# ============================================================================
# Linalg (3 tests)
# ============================================================================


class TestLinalg:
    """Linear algebra on GPU — tests 30-32."""

    def test_matmul(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        b_np = np.array([[5, 6], [7, 8]], dtype=np.float64)
        result = ins.matmul(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np @ b_np, rtol=1e-10)

    def test_det(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float32)
        try:
            result = ins.det(to_gpu(a_np))
            assert result.numpy().item() == pytest.approx(-2.0, abs=1e-3)
        except RuntimeError:
            pytest.skip("det requires cuBLAS")

    def test_inv(self):
        a_np = np.array([[4, 7], [2, 6]], dtype=np.float32)
        try:
            result = ins.inv(to_gpu(a_np))
            np.testing.assert_allclose(to_numpy(result), np.linalg.inv(a_np), atol=1e-3)
        except RuntimeError:
            pytest.skip("inv requires cuBLAS")


# ============================================================================
# FFT (2 tests)
# ============================================================================


class TestFFT:
    """FFT on GPU — tests 33-34."""

    def test_fft(self):
        x_np = np.array([1, 2, 3, 4], dtype=np.float64)
        result = ins.fft(to_gpu(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(to_numpy(result), expected, atol=1e-8)

    def test_ifft(self):
        x_np = np.array([1, 2, 3, 4], dtype=np.float64)
        fft_np = np.fft.fft(x_np)
        result = ins.ifft(to_gpu(fft_np.astype(np.complex128)))
        expected = np.fft.ifft(fft_np)
        np.testing.assert_allclose(to_numpy(result), expected, atol=1e-8)


# ============================================================================
# Signal (1 test)
# ============================================================================


class TestSignal:
    """Signal — test 35."""

    def test_hann(self):
        w = ins.signal.hann(16)
        assert w.numel == 16
        assert w.numpy()[0] == pytest.approx(0.0, abs=1e-5)
        assert w.numpy()[8] == pytest.approx(1.0, abs=1e-1)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
